rust/macros/pinned_drop.rs

```diff
@@ -0,0 +1,82 @@
+// SPDX-License-Identifier: GPL-2.0
+
+use proc_macro::{Delimiter, Group, Ident, Punct, Spacing, Span, TokenStream, TokenTree};
+
+pub(crate) fn pinned_drop(_args: TokenStream, input: TokenStream) -> TokenStream {
+    let mut toks = input.into_iter().collect::<Vec<_>>();
+    assert!(!toks.is_empty());
+    // ensure that we have an impl item
+    assert!(matches!(&toks[0], TokenTree::Ident(i) if i.to_string() == "impl"));
+    // ensure that we are implementing `PinnedDrop`
+    let mut nesting: usize = 0;
+    let mut pinned_drop_idx = None;
+    for (i, tt) in toks.iter().enumerate() {
+        match tt {
+            TokenTree::Punct(p) if p.as_char() == '<' => {
```

Comment:
Perhaps the angle bracket matching code can be extracted to a function and shared between both macros.

Reply:
This seems a bit difficult, because in the code for `pin_project` I am going through the generics list and I am searching for the identifiers. So that code seems a bit too specific to extract that and keep the functionality...

I can still create the function for use in other places though...